File: backend/src/app/schemas/valorizations.py
"""Valorization schemas."""

import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field

from app.models.enums import IndexType, ValorizationStatus
from app.schemas.common import ORMBaseSchema


class ValorizationCreate(BaseModel):
    """Request payload for creating a valorization."""

    contract_id: uuid.UUID
    year: int
    index_type: IndexType
    index_value: Decimal
    planned_date: date
    applied_date: date | None = None
    status: ValorizationStatus = ValorizationStatus.PENDING
    approved_by: uuid.UUID | None = None
    notes: str | None = None
    additional_data: dict = Field(default_factory=dict)


class ValorizationUpdate(BaseModel):
    """Request payload for partial update of a valorization."""

    contract_id: uuid.UUID | None = None
    year: int | None = None
    index_type: IndexType | None = None
    index_value: Decimal | None = None
    planned_date: date | None = None
    applied_date: date | None = None
    status: ValorizationStatus | None = None
    approved_by: uuid.UUID | None = None
    notes: str | None = None
    additional_data: dict | None = None


class ValorizationRead(ORMBaseSchema):
    """Valorization API response."""

    id: uuid.UUID
    contract_id: uuid.UUID
    year: int
    index_type: IndexType
    index_value: Decimal
    planned_date: date
    applied_date: date | None
    status: ValorizationStatus
    approved_by: uuid.UUID | None
    notes: str | None
    additional_data: dict
    created_by: uuid.UUID | None
    created_at: datetime
    updated_at: datetime


class ValorizationAutoItem(BaseModel):
    """Automation rule per contract."""

    contract_id: uuid.UUID
    index_type: IndexType | None = None
    index_value: Decimal | None = None


class ValorizationAutoRequest(BaseModel):
    """Request payload for auto-generating valorizations."""

    planned_date: date
    items: list[ValorizationAutoItem]


class ValorizationAutoSkipped(BaseModel):
    """Skipped contract during auto-generation."""

    contract_id: uuid.UUID
    reason: str


class ValorizationAutoResponse(BaseModel):
    """Auto-generation response summary."""

    planned_date: date
    year: int
    gus_value: Decimal | None
    created: list[ValorizationRead]
    skipped: list[ValorizationAutoSkipped]

File: frontend/src/hooks/valorizations.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'

import { apiClient } from '@/lib/axios'
import type {
  IndexType,
  Valorization,
  ValorizationCreate,
  ValorizationStatus,
  ValorizationUpdate,
} from '@/types/models'

const BASE = '/api/v1'

export interface ValorizationsFilters {
  contract_id?: string
  year?: number
  status?: ValorizationStatus
}

export interface ValorizationAutoItem {
  contract_id: string
  index_type?: IndexType
  index_value?: number | null
}

export interface ValorizationAutoRequest {
  planned_date: string
  items: ValorizationAutoItem[]
}

export interface ValorizationAutoSkipped {
  contract_id: string
  reason: string
}

export interface ValorizationAutoResponse {
  planned_date: string
  year: number
  gus_value?: number | null
  created: Valorization[]
  skipped: ValorizationAutoSkipped[]
}

export function useValorizations(filters?: ValorizationsFilters) {
  return useQuery({
    queryKey: ['valorizations', filters],
    queryFn: async () => {
      const { data } = await apiClient.get<Valorization[]>(`${BASE}/valorizations`, {
        params: filters,
      })
      return data
    },
  })
}

export function useCreateValorization() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (payload: ValorizationCreate) => {
      const { data } = await apiClient.post<Valorization>(`${BASE}/valorizations`, payload)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['valorizations'] })
    },
  })
}

export function useUpdateValorization() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({ id, payload }: { id: string; payload: ValorizationUpdate }) => {
      const { data } = await apiClient.patch<Valorization>(`${BASE}/valorizations/${id}`, payload)
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['valorizations'] })
    },
  })
}

export function useDeleteValorization() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (id: string) => {
      await apiClient.delete(`${BASE}/valorizations/${id}`)
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['valorizations'] })
    },
  })
}

export function useGenerateValorizations() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (payload: ValorizationAutoRequest) => {
      const { data } = await apiClient.post<ValorizationAutoResponse>(
        `${BASE}/valorizations/auto`,
        payload
      )
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['valorizations'] })
    },
  })
}
